src/java_bytecode/expr2java.cpp

```diff
@@ -237,23 +237,40 @@ std::string expr2javat::convert_constant(
     dest+='L';
     return dest;
   }
-  else if (src.type() == java_float_type()) {
-    ieee_floatt Fpn(to_constant_expr(src));
-    float F = Fpn.to_float();
-    std::stringstream buffer;
-    buffer << std::hexfloat;
-    buffer << F << 'f';
-    return(buffer.str());
-  }
-   else if (src.type() == java_double_type()) {
-    ieee_floatt Fpn(to_constant_expr(src));
-    double D = Fpn.to_double();
+  else if((src.type()==java_float_type()) ||
+          (src.type()==java_double_type()))
+  {
+    ieee_floatt ieee_repr(to_constant_expr(src));
+    if(ieee_repr.is_NaN() || ieee_repr.is_infinity())
+    {
+      std::string result;
+      if(ieee_repr.get_sign())
+        result="-";
+      else
+        result="+";
+      if(ieee_repr.is_NaN())
+        result+="NaN";
+      else
+        result+="inf";
+      return(result);
+    }
+
+    bool float_flag=(src.type()==java_float_type());
+    float f_num;
+    double d_num;
+    if(float_flag)
+      f_num=ieee_repr.to_float();
+    else
+      d_num=ieee_repr.to_double();
     std::stringstream buffer;
     buffer << std::hexfloat;
-    buffer << D;
+    if(float_flag)
```

Comment:
nitpick, explicit `bool` is not needed, `if(src.type()==java_float_type())` would be enough